src/library/autodj/autodjprocessor.cpp

```diff
@@ -817,61 +965,204 @@ void AutoDJProcessor::calculateTransition(DeckAttributes* pFromDeck,
-            pFromDeck->fadeDuration = m_transitionTime / fromTrackDuration;
+        break;
+    case TransitionMode::AlignIntroOutroStart:
+        // Use the outro or intro length for the transition time, whichever is
+        // shorter. If the outro is longer than the intro, cut off the end
+        // of the outro.
+        //
+        // In the diagrams below,
+        // - is part of a track outside the outro/intro,
+        // o is part of the outro
+        // i is part of the intro
+        // | marks the boundaries of the transition
+        //
+        // When outro > intro:
+        // ------|ooo|ooo
+        //       |iii|------
+        //
+        // When outro < intro:
+        // ------|ooo|
+        //       |iii|iii-----
+        //
+        // If only the outro or intro length is marked but not both, use the one
+        // that is marked for the transition time. If neither is marked, fall
+        // back to the transition time from the spinbox.
+        if (outroLength > 0 && introLength > 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            if (outroLength > introLength) {
+                // Cut off end of outro
+                pFromDeck->fadeEndPos = outroStart + introLength;
+            } else {
+                pFromDeck->fadeEndPos = outroEnd;
+            }
+            pToDeck->startPos = introStart;
+        } else if (outroLength > 0 && introLength <= 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            pFromDeck->fadeEndPos = outroEnd;
+            pToDeck->startPos = introStart;
+        } else if (introLength > 0 && outroLength <= 0) {
+            pFromDeck->fadeBeginPos = outroEnd - introLength;
+            pFromDeck->fadeEndPos = outroEnd;
+            pToDeck->startPos = introStart;
         } else {
-            pFromDeck->posThreshold = 1.0;
-            pFromDeck->fadeDuration = 0.0;
+            useFixedFadeTime(pFromDeck, pToDeck,
+                    getLastSoundPosition(pFromDeck),
+                    getFirstSoundPosition(pToDeck));
         }
+        break;
+    case TransitionMode::FixedSkipSilence:
+        useFixedFadeTime(pFromDeck,
+                pToDeck,
+                getLastSoundPosition(pFromDeck),
+                getFirstSoundPosition(pToDeck));
+        break;
+    case TransitionMode::FixedFullTrack:
+    default:
+        useFixedFadeTime(pFromDeck, pToDeck, fromTrackDuration, 0);
+    }
 
-        if (toTrackDuration > 0.0) {
-            // Transition time is negative here, so we need to add instead of subtract.
-            pToDeck->startPos = (toTrackIntroStart + m_transitionTime) / toTrackDuration;
-        } else {
-            pToDeck->startPos = 0.0;
+    // These are expected to be a fraction of the track length.
+    pFromDeck->fadeBeginPos /= fromTrackDuration;
+    pFromDeck->fadeEndPos /= fromTrackDuration;
+    pToDeck->startPos /= toTrackDuration;
+
+    pFromDeck->isFromDeck = true;
```

Comment:
I am afraid this is part of the bug I have described. We should get rid of this since it is possible to calculate the transition vice versa.

Reply:
I have implemented a solution to that bug that works with this explicit tracking of the "from deck": 
05351e6 I think reacting to the crossfader position changes is more robust and makes for easier to understand code than adding complexity to the already complicated playerPositionChanged function.